### ISSUE TITLE
Generating point clouds using a lidar model

### ISSUE THREAD

**Issue by eduardohenriquearnold** [facebookresearch/pytorch3d#218]
Given a set of meshes I would like to generate point clouds as if obtained from a rotating lidar sensor (i.e. using a lidar sensor model). 

One naive approach would be to use the depth buffer from the rasterizer and unproject the points, which is what I have been doing so far. However this misses the geometry of the scanned point cloud. Instead, I would like to implement a rotating lidar model using spherical coordinates to sample points at specific polar and azimuthal ray angles. 

Is ray tracing the optimal solution in this case? Or is there an easier way to implement this? 
I would like to work on a solution and contribute to the project.

**Comment by eduardohenriquearnold** [comment]
@nikhilaravi no, because that would yield a uniform sample over the surfaces. I am essentially doing the following:
```python
image_size = 64
rasterizer = MeshRasterizer(cameras=cameras, raster_settings=raster_settings)
frag = rasterizer(mesh)
zbuf = frag.zbuf

#Get correponding 3D PCL
u = torch.arange(image_size, dtype=torch.float32, device=device)/image_size
uu,vv = torch.meshgrid(u,u)
uvd = torch.cat([uu.reshape(-1,1), vv.reshape(-1,1), zbuf.reshape(-1,1)], dim=1)
xyz = cameras.unproject_points(uvd, world_coordinates=False)
```
However this does not follow a realistic rotating lidar model which sample points at specific azimuthal and polar angles in spherical coordinates. For example, the result I would like to observe for a torus and a sensor placed at a specific position would be similar to the image below (image from Blensor):
![](https://2.bp.blogspot.com/-uH0ZDjgWw1I/UpMpoc-pygI/AAAAAAAAACc/AJLEOCaTuEY/s1600/sim-scan.png)

I could summarise the problem as follows:
Given a unit vector with specified azimuthal and polar angles (spherical coordinates) at the sensor origin, find the minimum scalar such that this scalar multiplied by the unit vector intersects a mesh surface. We need to solve this problem for a grid of azimuthal and polar angles efficiently. 

My current idea is to find the intersection of the line resulting from the span of the unit vector with each face and return the one with the minimal distance, if it exists, using [this](https://stackoverflow.com/a/25704822) algorithm. Can I assume that all faces are composed of triangles? If not, do you recommend any specific algorithm to efficiently compute the intersection of a line and a face?

**Comment by gkioxari** [comment]
Hi @eduardohenriquearnold 
I am not sure I fully understand your task, but based on your great summary I can offer some suggestions. First, all of our meshes are assumed to have triangular faces (even if the input mesh is not triangular, we make it when we read it). We have tons of functionality to compute intersection of points with faces both in 2D and in 3D. These are used internally for rasterization and also for some loss functions (like `point_mesh_distance`). These basic functions live in `pytorch3d/csrc/utils/geometry_utils.cuh` for CUDA and `pytorch3d/csrc/utils/geometry_utils.h` for C++ (we are in the process of adding more for C++ as we have been focusing mostly on CUDA). 

For example,

https://github.com/facebookresearch/pytorch3d/blob/e3819a49dfa855de1a7c99c0583fb69f9bdad75b/pytorch3d/csrc/utils/geometry_utils.cuh#L535

computes the distance of a point wrt a triangle in 3D.  And its  backward

https://github.com/facebookresearch/pytorch3d/blob/e3819a49dfa855de1a7c99c0583fb69f9bdad75b/pytorch3d/csrc/utils/geometry_utils.cuh#L582

Poke more around in those files to find more functionality. 

Let me know if that is helpful and offers a solution to your problem.

**Comment by eduardohenriquearnold** [comment]
Hi @gkioxari. Firstly, thanks for the prompt and thorough response. I am sorry if I couldn't be clearer regarding the problem, but the information you provided can definitely help me! I will do some experiments and get back to you soon. Thanks!

**Comment by gkioxari** [comment]
Sounds good! Let us know and if you find any useful function abstractions that might be useful for others, we are happy to incorporate them into the codebase.

**Comment by gkioxari** [comment]
I will close this. If any issues arise feel free to re-open.
